test/setup-kubernetes.sh

```diff
@@ -267,7 +265,17 @@ cat ${kubeadm_config_file}
 # 5.0.9-301.fc30.x86_64).
 kubeadm_args="$kubeadm_args --ignore-preflight-errors=SystemVerification"
 
-kubeadm_args_init="$kubeadm_args_init --config=$kubeadm_config_file"
+# Migrate to newer config version incase if we are using outdated config schema.
+# `kubeadm config migrate` vanishing the 'KubeletConfiguration' while doing this.
```

Comment:
Isn't that a bug? The config was valid, why should something just disappear?

Presumably this is about the `kubelet.config.k8s.io/v1beta1`, right? That's still the latest version in 1.20, so what is out-dated?

What I am looking for is a clear explanation why this workaround is needed and when we can remove it again. Ideally we should also file an issue as reminder.

Reply:
> Isn't that a bug? The config was valid, why should something just disappear?

Might not be, I am not sure how config migrate is supposed to work with kubelet config or not ,as it is not part of `kubadm.k8s.io` API.

> 
> Presumably this is about the `kubelet.config.k8s.io/v1beta1`, right? That's still the latest version in 1.20, so what is out-dated?

Isn't it v1beta2, according to https://github.com/kubernetes/kubernetes/commit/3a050344d0e6475ec7ead775f3f1db32bc837c83?
> 
> What I am looking for is a clear explanation why this workaround is needed and when we can remove it again. Ideally we should also file an issue as reminder.

I am myself not sure if this is a workaround or not.

Reply:
> > Isn't that a bug? The config was valid, why should something just disappear?
> 
> Might not be, I am not sure how config migrate is supposed to work with kubelet config or not ,as it is not part of `kubadm.k8s.io` API.

If in doubt, file an upstream bug and clarify it.

> > Presumably this is about the `kubelet.config.k8s.io/v1beta1`, right? That's still the latest version in 1.20, so what is out-dated?
>
> Isn't it v1beta2, according to [kubernetes/kubernetes@3a05034](https://github.com/kubernetes/kubernetes/commit/3a050344d0e6475ec7ead775f3f1db32bc837c83)?

That v1beta2 is for kubeadm. I was talking about `kubelet.config.k8s.io/v1beta1` (= https://pkg.go.dev/k8s.io/kubelet/config) because that is the one that gets the special treatment.

> > What I am looking for is a clear explanation why this workaround is needed and when we can remove it again. Ideally we should also file an issue as reminder.
> 
> I am myself not sure if this is a workaround or not.

What we should do is generate a kubeadm config that works for the `kubeadm init` invocation. We shouldn't need `kubeadm migrate` - that just makes it harder to understand what's going on, in particular when we then restore parts that got lost during it.

Why is that not possible? I haven't tried it and it is not mentioned in the PR, so I don't know.

Reply:
> What we should do is generate a kubeadm config that works for the `kubeadm init` invocation

You mean write k8s version specific kubeadm configuration?

Reply:
> We shouldn't need `kubeadm migrate` - that just makes it harder to understand what's going on, in particular when we then restore parts that got lost during it.

You are right, This is causing the issue of misbehaving the kube-proxy, `kubeadm config migrate` only migrates the API configurations which it could validate, that is under APIs `kubeadm.k8s.io` other it just ignores, Because of this the feature-gates we set in the kubeproxy configuration is not passed to kube-proxy and causing the issue. I will revert this.

Reply:
Yes, or at least make the creation of that config a bit more flexible with k8s version checks.

I don't know how the configuration for 1.20 has to be different, so it's hard to tell without further information which approach is better.

Reply:
I had no problems bringing up 1.20 without this "kubeadm migrate". I removed it with the patch below and "make start" worked.

```
diff --git a/test/setup-kubernetes.sh b/test/setup-kubernetes.sh
index e10dec094..574324258 100755
--- a/test/setup-kubernetes.sh
+++ b/test/setup-kubernetes.sh
@@ -250,6 +250,8 @@ kubernetesVersion: $k8sversion
 cat >${kubeadm_config_file} <<EOF
 $kubeadm_config_init
 ---
+$kubeadm_config_kubelet
+---
 $kubeadm_config_cluster
 ---
 $kubeadm_config_proxy
@@ -265,17 +267,7 @@ cat ${kubeadm_config_file}
 # 5.0.9-301.fc30.x86_64).
 kubeadm_args="$kubeadm_args --ignore-preflight-errors=SystemVerification"
 
-# Migrate to newer config version incase if we are using outdated config schema.
-# `kubeadm config migrate` vanishing the 'KubeletConfiguration' while doing this.
-# So, append to post-migarted configuraiton.
-kubeadm_config_new_file="/tmp/kubeadm-config-new.yaml"
-kubeadm config migrate --old-config $kubeadm_config_file --new-config $kubeadm_config_new_file
-cat >>${kubeadm_config_new_file} <<EOF
----
-$kubeadm_config_kubelet
-EOF
-
-kubeadm_args_init="$kubeadm_args_init --config=$kubeadm_config_new_file"
+kubeadm_args_init="$kubeadm_args_init --config=$kubeadm_config_file"
 sudo kubeadm init $kubeadm_args $kubeadm_args_init || (
     set +e
     # Dump some information that might explain the failure.
```